New validator client using REST API.

cheatfate opened the pull request.

cheatfate committed:
Initial commit.

cheatfate committed:
Exporting getConfig().

cheatfate committed:
Add beacon node checking procedures.

cheatfate committed:
Post rebase fixes.

cheatfate committed:
Use runSlotLoop() from nimbus_beacon_node.
Fallback implementation.
Fixes for ETH2 REST serialization.

cheatfate committed:
Add beacon_clock.durationToNextSlot().
Move type declarations from beacon_rest_api to json_rest_serialization.
Fix seq[ValidatorIndex] serialization.
Refactor ValidatorPool and add some utility procedures.
Create separate version of validator_client.

cheatfate committed:
Post-rebase fixes.
Remove CookedPubKey from validator_pool.nim.

cheatfate committed:
Now we should be able to produce attestations and aggregate and proofs.
But its not working yet.

cheatfate committed:
Debugging attestation sending.

cheatfate committed:
Add durationToNextAttestation.
Optimize some debug logs.
Fix aggregation_bits encoding.
Bump chronos/presto.

cheatfate committed:
Its alive.

cheatfate committed:
Fixes for launch_local_testnet script.
Bump chronos.

tersec commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -1017,75 +1017,75 @@ proc onSlotStart(
-    curSlot = wallSlot
-    nextSlot = wallSlot + 1
-    timeToNextSlot = saturate(node.beaconClock.fromNow(nextSlot))
+# proc runSlotLoop(node: BeaconNode, startTime: BeaconTime) {.async.} =
```
It looks like this got moved to `nimbus_binary_common`. Is there still a reason to keep it extant but commented out, rather than removing the code from `nimbus_beacon_node` entirely, e.g., is it tentative or experimental?

cheatfate replied:
@tersec this code is not removed, but moved to `nimbus_beacon_node` to share it between `nimbus_beacon_node` and `nimbus_validator_client`.

cheatfate replied:
But of course i can copy-paste same code from `nimbus_beacon_node` to `nimbus_validator_client` and add comments, that if logic is going to be changed it should be changed in `validator_client` source too.

cheatfate replied:
Removed

cheatfate committed:
Switch client API to not use `/api` prefix.

cheatfate committed:
Post-rebase adjustments.

cheatfate committed:
Fix endpoint for publishBlock().

cheatfate committed:
Add CONFIG_NAME.
Add more checks to ensure that beacon_node is compatible.

arnetheduck commented on beacon_chain/beacon_clock.nim:
```diff
@@ -92,6 +92,26 @@ proc now*(c: BeaconClock): BeaconTime =
   ## Current time, in slots - this may end up being less than GENESIS_SLOT(!)
   toBeaconTime(c, getTime())
 
+proc durationToNextSlot*(c: BeaconClock): Duration =
```
reuse `fromNow`?

cheatfate replied:
Done.

arnetheduck commented on beacon_chain/nimbus_validator_client.nim:
```diff
@@ -1,341 +1,201 @@
+      error "Duplicate validator's key found", validator_pubkey = pubkey
+      return false
+    else:
+      duplicates.add(pubkey)
```
make the duplicate check part of `addLocalValidator` instead?

cheatfate replied:
This check should be done on `validator_client` because it is supposed to manage validator keys (add/remove). But for `beacon_node` itself it doesn't have any meaning because you can't create two files/folders with equal name.

arnetheduck commented on beacon_chain/nimbus_validator_client.nim:
```diff
@@ -1,341 +1,201 @@
+      vc.attachedValidators.addLocalValidator(key)
+  return true
+
+proc initClock*(vc: ValidatorClientRef): Future[BeaconClock] {.async.} =
```
is it possible to compare the time of the beacon node with local wall time somehow, through some request? ie we'd have a tolerance of maybe .. 500ms or something

cheatfate replied:
I can combine calls to `/eth/v1/beacon/genesis` to get genesis time and `/eth/v1/node/syncing` to get latest head. But accuracy would be something near `12.seconds`.

arnetheduck replied:
this almost merits a new RPC call, something to think about

arnetheduck commented on beacon_chain/validator_client/attestation_service.nim:
```diff
@@ -0,0 +1,185 @@
+    let sres = vc.getCurrentSlot()
+    if sres.isSome():
+      let currentSlot = sres.get()
+      asyncSpawn service.spawnAttestationTasks(currentSlot)
```
if tasks take longer than a slot, this will keep spawning lots of tasks - why not `await` here?

cheatfate replied:
Fixed.

arnetheduck commented on beacon_chain/validator_client/attestation_service.nim:
```diff
@@ -0,0 +1,185 @@
+    await sleepAsync(sleepTime)
+
+proc init*(t: typedesc[AttestationServiceRef],
+           vc: ValidatorClientRef): Future[AttestationServiceRef] {.async.} =
```
hm, "just-in-case"-async? I generally think such constructs make the code more complex unnecessarily - this is not a library that needs to prepare the api for future use cases

cheatfate replied:
Of course but `AttestationServiceRef` is just one of the services, so to avoid code duplication for managing services i decided to use async version, because some services require to perform some activities before start.

arnetheduck commented on beacon_chain/validator_client/block_service.nim:
```diff
@@ -0,0 +1,103 @@
+    vc = service.client
+    validator = vc.attachedValidators.validators[pubkey]
+    genesisRoot = vc.beaconGenesis.genesis_validators_root
+    graffiti =
```
```suggestion
    graffiti = vc.conf.graffiti.get(defaultGraffitiBytes())
```

cheatfate replied:
Fixed.

arnetheduck commented on beacon_chain/validator_client/block_service.nim:
```diff
@@ -0,0 +1,103 @@
+    debug "Randao", randao_reveal = randaoReveal
+    let beaconBlock = await vc.produceBlock(slot, randaoReveal, graffiti)
+    debug "Block", non_signed_block = beaconBlock
+    var signedBlock = SignedBeaconBlock(message: beaconBlock,
```
this instance can be created further down, when we already have the signature

cheatfate replied:
Fixed.

arnetheduck commented on beacon_chain/validator_client/common.nim:
```diff
@@ -0,0 +1,192 @@
+    endpoint*: string
+    config*: Option[RestConfig]
+    ident*: Option[string]
+    genesis*: Option[RestBeaconGenesis]
```
for servers to be valid/useful, they have to have a genesis, no?

cheatfate replied:
`Option` here is used to just confirm that node do not returned `genesis`.

arnetheduck commented on beacon_chain/validator_client/duties_service.nim:
```diff
@@ -0,0 +1,352 @@
+            proposers_count = len(initialProposers)
+      # Notify the `BlockServiceRef` for any proposals that we have in our
+      # cache.
+      await vc.notifyBlockProductionService(currentSlot,
```
this queue looks redundant and potentially problematic - in particular, if there is a reorg while the duty sits in the queue, it'll be stale

cheatfate replied:
Queue was removed.

arnetheduck commented on beacon_chain/rpc/beacon_rest_api.nim:
```diff
@@ -946,3 +927,38 @@ proc installBeaconApiHandlers*(router: var RestRouter, node: BeaconNode) =
+     meth: MethodGet.}
+  ## https://ethereum.github.io/eth2.0-APIs/#/Beacon/getStateFork
+
+proc publishBlock*(body: SignedBeaconBlock): RestPlainResponse {.
```
for generic code, it might be easier if `RestResponse[void]` is used for empty responses

cheatfate replied:
`RestResponse[void]` will attempt to decode REST server response to `void` type... While `RestPlainResponse` will just return `seq[byte]` which can be decoded at runtime.

cheatfate committed:
Add beacon committee subscription support to validator_client.

cheatfate committed:
Fix stacktrace should be an array of strings.
Fix committee subscriptions should not be `data` keyed.

cheatfate committed:
Log duration to next block proposal.

cheatfate committed:
Fix beacon_node_status import.

cheatfate committed:
Use jsonMsgResponse() instead of jsonError().

cheatfate committed:
Fix graffityBytes usage.
Remove unnecessary `await`.
Adjust creation of SignedBlock instance.
Remove legacy files.

cheatfate committed:
Rework durationToNextSlot() and durationToNextEpoch() to use `fromNow`.

arnetheduck commented on beacon_chain/validator_client/fork_service.nim:
```diff
@@ -0,0 +1,34 @@
+  return res
+
+proc start*(service: ForkServiceRef) =
+  service.lifeFut = mainLoop(service)
```
using a future this way means that if an exception breaks the service, we'll not notice

cheatfate replied:
Fixed.

arnetheduck commented on beacon_chain/validator_client/attestation_service.nim:
```diff
@@ -0,0 +1,185 @@
+
+        let attestation = await validator.produceAndSignAttestation(ad,
+          int(duty.committee_length), Natural(duty.validator_committee_index),
+          vc.fork.get(), vc.beaconGenesis.genesis_validators_root)
```
this is a race condition: because `vc.fork` is being updated by a separate task, it's possible that it changes while this loop is running - the fork depends on the slot we're attesting for, so when attesting for slot N, we should lock down a fork for that slot

cheatfate replied:
There is no more loop present, and all calls to `produceAndSignAttestation` are produced concurrently, with single `vc.fork`. So no more race condition.

arnetheduck commented on beacon_chain/validator_client/duties_service.nim:
```diff
@@ -0,0 +1,419 @@
+    vc.pruneAttesterDuties(currentEpoch)
+
+proc getBlockProposers*(vc: ValidatorClientRef,
+                        slot: Slot): HashSet[ValidatorPubKey] =
```
there can only be one proposer per slot - why the hashset?

cheatfate replied:
Even if there one proposer per slot, https://ethereum.github.io/eth2.0-APIs/#/Validator/getProposerDuties could return any mount of proposers per slot.

arnetheduck commented on beacon_chain/validator_client/duties_service.nim:
```diff
@@ -0,0 +1,419 @@
+proc validatorIndexLoop(service: DutiesServiceRef) {.async.} =
+  let vc = service.client
+  while true:
+    await vc.pollForValidatorIndices()
```
when all attached validatorindices are known, does this loop continue polling?

cheatfate replied:
Of course, because new validators could be added, but if all validator indices are already known, call to `vc.pollForValidatorIndices()` is NOP call.

cheatfate committed:
Fix race condition for block proposal and attestations for same slot.
Fix local_testnet script to properly kill tasks on Windows.
Bump chronos and nim-http-tools, to allow connections to infura.io (basic auth).

cheatfate committed:
Catch services errors.
Improve performance of local_testnet.sh script on Windows.
Fix race condition when attestation producing.

cheatfate committed:
Post-rebase fixes.

cheatfate committed:
Bump chronos and presto.

cheatfate committed:
Calculate block publishing delay.
Fix pkill in one more place.

arnetheduck commented on beacon_chain/validator_client/attestation_service.nim:
```diff
@@ -0,0 +1,205 @@
+
+  let count = len(attestations)
+  if count > 0:
+    let res = await vc.submitPoolAttestations(attestations)
```
interesting that it makes sense to submit attestations together - I would think that in general it's better to produce and send them one by one since  they're completely independent

arnetheduck replied:
basically, this means that we wait until all signatures are done before starting the broadcast, then burst the broadcast in one go - this has two issues: the first attestation is delayed until the last attestation is done decreasing the chances it gets included and we cause a bandwidth burst when sending off all these attestations at once instead of staggering them

cheatfate replied:
Fixed, now all the attestations are working in tasks.

arnetheduck commented on beacon_chain/validator_client/common.nim:
```diff
@@ -0,0 +1,236 @@
+  var minimumDuration = InfiniteDuration
+  var currentSlotTime = Duration(slot.toBeaconTime())
+  let currentEpoch = slot.epoch()
+  for epoch in [currentEpoch, currentEpoch + 1'u64]:
```
if we find a proposal in the current epoch, we should not ask for the next epoch

cheatfate replied:
Could you explain why?

arnetheduck replied:
well, the duties in the next epoch are guaranteed to be for slots that come later than those of the current epoch, so they cant be earlier meaning they won't have a lower timeleft

cheatfate replied:
Fixed.

cheatfate committed:
Add error handling and timeouts to firstSuccess() template.
Add onceToAll() template.
Add checkNodes() procedure.
Refactor firstSuccess() template.
Add error checking to api.nim calls.

cheatfate committed:
Deprecated usage onceToAll() for better stability.
Address comment and send attestations asap.

cheatfate committed:
Avoid unnecessary loop when calculating minimal duration.
